src/box/lua/config/applier/credentials.lua

```diff
@@ -277,6 +277,101 @@ local function privileges_add_defaults(name, role_or_user, intermediate)
+-- after object creation. Actions are listed in the following format:
+-- [obj_type][obj_name] = {{
+--     role_or_user,
+--     grant_or_revoke,
+--     user_or_role_name,
+--     privileges
+--   }, {
+--     ...
+--   },
+-- }
+local actions_on_creation = {
+    ['space'] = {},
+    ['function'] = {},
+    ['sequence'] = {},
+}
+
+-- Info about objects is stored inside system spaces
+-- box.space._space/_func/_sequence. Translate these names
+-- to obj_type values of space/function/sequence.
+local function translate_obj_type(obj_type)
+    if obj_type == '_space' then
+        return 'space'
+    elseif obj_type == '_func' then
+        return 'function'
+    elseif obj_type == '_sequence' then
+        return 'sequence'
+    end
+end
+
+-- Iterate through events and apply postponed action for newly created objects.
+local function on_commit_trigger(iterator)
+    log.debug('on_commit_trigger started')
+    for _, _, new_space, space_id in iterator() do
+        local obj_type = box.space[space_id].name
+        obj_type = translate_obj_type(obj_type)
+
+        -- Check if the event is for space/function/sequence.
+        if obj_type == nil then
+            goto skip
+        end
+
+        local obj_name = new_space.name
+        local obj_info = actions_on_creation[obj_type][obj_name] or {}
+
+        -- Apply all postponed actions for the object.
+        for _, info in ipairs(obj_info) do
+            local role_or_user, grant_or_revoke, name, privs = unpack(info)
+            -- There is a case when two objects have the same perms for the same
+            -- user and one is renamed to another, then the grant happens twice,
+            -- thus use if_not_exists = true.
+            box.schema[role_or_user][grant_or_revoke](name, privs, obj_type,
+                                            obj_name, {if_not_exists = true})
+
+            log.debug('credentials.apply: box.schema.%s.%s(%q, %q, %q, %q)',
+                      role_or_user, grant_or_revoke, name, privs,
+                      obj_type, obj_name)
+       end
+       actions_on_creation[obj_type][obj_name] = nil
+       ::skip::
+    end
+    log.debug('on_commit_trigger finished')
+end
+
+-- Check that event is space/function/sequence creation and set
+-- on_commit_trigger for the transaction.
+local function on_replace_trigger(_, new_space, obj_type, request_type)
+    log.debug('on_replace_trigger started')
+    obj_type = translate_obj_type(obj_type)
+
+    -- Check if the event is object creation or rename
+    -- and object is space/function/sequence.
+    if (request_type ~= 'INSERT' and request_type ~= 'UPDATE') or
+            obj_type == nil then
+        return
+    end
+
+    -- Check that there actions postponed for the object.
+    local obj_name = new_space.name
+    local obj_info = actions_on_creation[obj_type][obj_name] or {}
+    if not next(obj_info) then
+        return
+    end
```

Comment:
If we rename a space, we should revoke all the privileges that we given by the provious configuration. IOW, we should finish in the state that is equivalent to one after `:reload()` with the new config.

Reply:
If we reuse the config -> box synchronization code on any trigger wake up, this problem should gone.

Reply:
> If we rename a space

Generally speaking, we don't rename spaces. And when an object gets changed by users hand by `box.*` command, it isn't our responsibility to change configuration afterwards. We explicitly agreed that all the privileges granted after the applier are not to be revoked. I am not sure that in this case it is a good idea either.
Another question is, what do we do if the renaming happens twice? Do we sync privileges each time or just once?

Reply:
We've discussed it verbally and agreed to revoke excessive privileges on rename. It is done with the brand-new per-object sync that is implemented in the `TO_SQUASH` commit.